### ISSUE TITLE
Race condition in CachedDict.__getitem__

### ISSUE THREAD

**Issue by adamchainz** [YPlan/django-modeldict#18]
YPlan/gargoyle#50 reports a race condition that exists inside `CachedDict.__getitem__`, when using a threaded server that can touch the `CachedDict` instance in parallel. Stack trace:

```
  File "/var/www/env/local/lib/python2.7/site-packages/gargoyle/manager.py", line 59, in is_active
    switch = self[key]
  File "/var/www/env/local/lib/python2.7/site-packages/gargoyle/manager.py", line 35, in __getitem__
    return SwitchProxy(self, super(SwitchManager, self).__getitem__(key))
  File "/var/www/env/local/lib/python2.7/site-packages/modeldict/base.py", line 26, in __getitem__
    return self._local_cache[key]
TypeError: 'NoneType' object has no attribute '__getitem__'
```

Probably `_populate` should be updated to be race-condition free and return the state of the local cache.

**Comment by kmorrison** [comment]
Apologies for the late reply on this. We were using a gevented ioloop when we were seeing this error. We have since stopped using gevent and have moved to 1 request per process, and haven't seen this error since. However, due to the infrequent nature I've the error, I am hesitant to draw conclusions thus far. I will post further if this error crops up again in a non-threaded environment.

**Comment by adamchainz** [comment]
No worries. I guess this issue is on freeze until it reappears...

**Comment by adamchainz** [comment]
#20 might have fixed this

**Comment by adamchainz** [comment]
or #25 might have

**Comment by Jaidan** [comment]
I can confirm this is not fixed currently.  It appears that one thread can cause a reset while another thread checks the value.  Several approaches to fixing this:

`_populate` could return an object safe to use for the rest of the thread.
`_local_cache` could be improved to be a read-through cache to the remote cache and reset behavior would empty the local dictionary instead of setting to `None` and `__getitem__` to read and populate.

**Comment by Jaidan** [comment]
@adamchainz We opened a PR that we're using that should address fix the thread safety issue.

**Comment by adamchainz** [comment]
Thanks for the bump, the PR slipped my radar. Reviewing now.

**Comment by adamchainz** [comment]
Fixed by #40
